src/requestManager.cpp

```diff
@@ -394,67 +400,77 @@ void CRequestManager::SendRequests()
+      if (IsChainNearlySyncd())
+          MIN_BLK_REQUEST_RETRY_INTERVAL = 5*1000*1000;
+      else
+          MIN_BLK_REQUEST_RETRY_INTERVAL = 30*1000*1000;
 
       if (now-item.lastRequestTime > MIN_BLK_REQUEST_RETRY_INTERVAL)  // if never requested then lastRequestTime==0 so this will always be true
 	{
           if (!item.availableFrom.empty())
 	    {
 	      CNodeRequestData next;
-              while ((!item.availableFrom.empty())&&(next.node == NULL)) // Go thru the availableFrom list, looking for the first node that isn't disconnected
+              while (!item.availableFrom.empty() && (next.node == NULL)) // Go thru the availableFrom list, looking for the first node that isn't disconnected
                 {
   	        next = item.availableFrom.front();  // Grab the next location where we can find this object.
                 item.availableFrom.pop_front();
                 if (next.node != NULL)
                   {
-		    if (next.node->fDisconnect)  // Node was disconnected so we can't request from it
+                    // Do not request from this node if it was disconnected or the node pingtime is far beyond acceptable.
+                    // We only check pingtime during IBD because we don't want to lock vNodes too often and when the chain is syncd, waiting
+                    // just 5 seconds for a timeout is not an issue, however waiting for a slow node during IBD can really slow down the process.
+                    //   TODO: Eventually when we move away from vNodes or have a different mechanism for tracking ping times we can include
+                    //   this filtering in all our requests for blocks and transactions.
+		    if (next.node->fDisconnect || (!IsChainNearlySyncd() && !IsNodePingAcceptable(next.node)))
 		      {
-                      LOCK(cs_vNodes);
-                      LogPrint("req", "ReqMgr: %s removed ref to %d count %d (disconnect).\n",item.obj.ToString(), next.node->GetId(), next.node->GetRefCount());
-                      next.node->Release();
-                      next.node = NULL; // force the loop to get another node            
+                        LOCK(cs_vNodes);
+                        LogPrint("req", "ReqMgr: %s removed ref to %d count %d (disconnect).\n", item.obj.ToString(), next.node->GetId(), next.node->GetRefCount());
+                        next.node->Release();
+                        next.node = NULL; // force the loop to get another node            
 		      }
 		  }
 	        }
 
 	      if (next.node != NULL )
 		{
-		  if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
+                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
+                  //     The chain must be almost syncd and traffic shaping must not be turned on
+		  if (item.lastRequestTime && !IsTrafficShapingEnabled())
```

Comment:
why no rerequest if traffic shaping?  sure shaping could slow down the receipt of blocks, but clearly blocks could still be dropped if you are shaping or not.  It seems like we have an interaction with traffic shaping that could be problematic -- that is, there exists pathological configurations where your own traffic shaping causes block rerequests.  However turning off rerequesting if traffic shaping is turned on will cause runtime issues for drops during normal configuration.  

I think that the predicate we are looking for here is more like "ReceiveBufferRecentlyEmpty()"